### PR TITLE
Add `assign_core` benchmark to bulk assigner

### PR DESCRIPTION
**Pull request by BradleyOlson64**
When building benchmark code I found that previous changes to the `build` function in builder only worked for tests and not for benchmarks. The fix I thought of was to have one version of the function for tests which accommodates the mock assigner and another for benchmarks that accommodates whichever assigner a particular runtime is using.

There's likely a better solution, so I'd be very open to hearing it!

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4426052

File: polkadot/runtime/parachains/src/assigner_bulk/mod.rs [hunk 1]
```diff
@@ -45,15 +46,21 @@ use sp_std::prelude::*;
...
 
 /// A weight info that is only suitable for testing.
 pub struct TestWeightInfo;
 
 /// Fraction expressed as a nominator with an assumed denominator of 57,600.
 pub type PartsOf57600 = u16;
 
-impl WeightInfo for TestWeightInfo {}
+impl WeightInfo for TestWeightInfo {
+	fn assign_core() -> Weight {
```
Comment (eskimor) [inline polkadot/runtime/parachains/src/assigner_bulk/mod.rs:60]: This should be parameterized at least on the number of assignments. Are there other variables (parameters or state) that could influence the weight?

**Review by eskimor — COMMENTED** [review]
Thanks Bradley! Are you aware of @antonva 's PR [here](https://github.com/paritytech/polkadot-sdk/pull/2419)? We moved the extrinsic to its own "interface" pallet.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4449613

**Comment by eskimor** [comment]
> When building benchmark code I found that previous changes to the build function in builder only worked for tests and not for benchmarks

I fixed that by moving all the code that dependent on mock assigner to the place where it is actually known that mock assigner is in use. (The test that instantiates the runtime.)

The builder itself is assigner agnostic. (Parameterized over T)

File: polkadot/runtime/parachains/src/assigner_bulk/mod.rs (hunk 1 above)
Comment (BradleyOlson64) [reply polkadot/runtime/parachains/src/assigner_bulk/mod.rs]: Done! 

It doesn't seem like there are any other parameter or state influences, unless you count parameters that cause error cases. In error cases the resource costs are lower, so idk that it makes sense to account for them when benchmarking for the worst case.

All state accesses happen once each and are O(1)

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4485673

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4486255

File: polkadot/runtime/rococo/src/lib.rs [hunk 2]
```diff
@@ -998,6 +998,7 @@ impl coretime::Config for Runtime {
 	type Currency = Balances;
 	//type WeightInfo = weights::runtime_coretime::WeightInfo<Runtime>;
 	type ExternalBrokerOrigin = EnsureXcm<OnlyBroker>;
+	type WeightInfo = runtime_common::coretime::TestWeightInfo;
```
Comment (eskimor) [inline polkadot/runtime/rococo/src/lib.rs:1001]: We should use the benchmark result here.

**Review by eskimor — COMMENTED** [review]
Looks good @BradleyOlson64 ! Can we have the benchmarks run and use them in Rococo?

File: polkadot/runtime/rococo/src/lib.rs (hunk 2 above)
Comment (BradleyOlson64) [reply polkadot/runtime/rococo/src/lib.rs:1001]: Wasn't sure if it was okay to add the benchmark to Rococo yet. I'll do that!

**Comment by BradleyOlson64** [comment]
bot bench polkadot-pallet --runtime=rococo --pallet=runtime_common::coretime

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4528586

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4529004

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4548650

**Comment by BradleyOlson64** [comment]
bot bench polkadot-pallet --runtime=rococo --pallet=runtime_common::coretime

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4550815

File: polkadot/runtime/common/src/coretime/benchmarking.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+mod benchmarks {
+	use super::*;
+	#[benchmark]
+	fn assign_core(s: Linear<0, MAX_ASSIGNMENTS_PER_SCHEDULE>) {
+		// Setup
+		let caller: <T as frame_system::Config>::AccountId = whitelisted_caller();
+		// TODO: This value is copied from BROKER_ID in rococo runtime constants.
+		// Perhaps provide it in a more future proof way if worth the extra 
+		// dependency.
+		let broker_id: u32 = 1004;
```
Comment (eskimor) [inline polkadot/runtime/common/src/coretime/benchmarking.rs:39]: I see three options to resolve this:

1. Move the benchmark to the rococo runtime (and then we copy it to each chain) - not ideal. It is not that much code, but could still evolve.
2. Configure the [origin](https://github.com/paritytech/polkadot-sdk/blob/b9421f376f4db4b9fb18dcea295448776a04cf52/polkadot/runtime/common/src/coretime/mod.rs#L80) differently when benchmarks are enabled similar to what is done [here](https://github.com/paritytech/polkadot-sdk/blob/b9421f376f4db4b9fb18dcea295448776a04cf52/polkadot/runtime/rococo/src/lib.rs#L1286).
3. Avoid the origin check all together when doing benchmarks. Pretty sure it does not add any weight.

Comment (eskimor) [reply polkadot/runtime/common/src/coretime/benchmarking.rs:39]: The way it is now, will break because the broker chain will almost certainly have different ids on different networks.

File: polkadot/runtime/parachains/src/builder.rs [hunk 4]
```diff
@@ -702,11 +704,8 @@ impl<T: paras_inherent::Config> BenchBuilder<T> {
 			.map(|i| {
 				let AssignmentProviderConfig { ttl, .. } =
 					scheduler::Pallet::<T>::assignment_provider_config(CoreIndex(i));
-				// Load an assignment into provider so that one is present to pop
-				let assignment = <T as scheduler::Config>::AssignmentProvider::get_mock_assignment(
-					CoreIndex(i),
-					ParaId::from(i),
-				);
+				let assignment =
```
Comment (eskimor) [inline polkadot/runtime/parachains/src/builder.rs:707]: Why was this changed? And how is it supposed to work? We don't know the underlying assigner, hence we can't know whether the pop_assignment_for_core will yield any assignment. Depending on configuration this unwrap will fail.

Comment (BradleyOlson64) [reply polkadot/runtime/parachains/src/builder.rs:707]: This is from a Git mishap. I'll clean it up.

**Comment by BradleyOlson64** [comment]
bot bench polkadot-pallet --runtime=rococo --pallet=runtime_common::coretime

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4564568

**Comment by BradleyOlson64** [comment]
bot bench polkadot-pallet --runtime=rococo --pallet=runtime_common::coretime

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4564726

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4564724

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4564725

**Comment by BradleyOlson64** [comment]
bot bench polkadot-pallet --runtime=rococo --pallet=runtime_common::coretime

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4564881

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4564880

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4564879

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4565270

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4565271

**Review by eskimor — COMMENTED** [review]
Looks good. CI still complains about some warnings.

File: polkadot/runtime/parachains/src/assigner_parachains.rs [hunk 5]
```diff
@@ -17,10 +17,6 @@
 //! The bulk (parachain slot auction) blockspace assignment provider.
 //! This provider is tightly coupled with the configuration and paras modules.
 
-mod mock_helpers;
```
[the remark this answers is not in the document]
Comment (BradleyOlson64) [reply polkadot/runtime/parachains/src/assigner_parachains.rs]: Thanks for catching that. 

I remembered from earlier in my bulk assigner work that there were no `tests` or `mock_helpers` files. But since then those files were added.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4575537

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4575538

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4575685
